File: integrations/agntcy-a2a-slimrpc/fixtures/dotnet/Observation.cs
// Copyright AGNTCY Contributors (https://github.com/agntcy)
// SPDX-License-Identifier: Apache-2.0

using System.Text;
using A2A;

namespace Csit;

/// <summary>
/// Parseable view of a SendMessage result, mirroring the Go probe's observation struct.
/// Emitted as CSIT_SLIM_* KEY=value lines consumed by matrix_test.go.
/// </summary>
internal sealed class Observation
{
    public string Kind { get; set; } = "unknown"; // "task" | "message" | "unknown"
    public string State { get; set; } = "";        // TASK_STATE_* token; empty for a bare message
    public bool ArtifactPresent { get; set; }
    public string Text { get; set; } = "";
    public int StreamEvents { get; set; }

    public static Observation FromResponse(SendMessageResponse resp) => resp.PayloadCase switch
    {
        SendMessageResponseCase.Message => new Observation { Kind = "message", Text = MessageText(resp.Message) },
        SendMessageResponseCase.Task => FromTask(resp.Task!),
        _ => new Observation { Kind = "unknown" },
    };

    public static Observation FromTask(AgentTask task)
    {
        var (text, present) = TaskArtifactText(task);
        return new Observation
        {
            Kind = "task",
            State = StateToken(task.Status?.State),
            ArtifactPresent = present,
            Text = text,
        };
    }

    /// <summary>Fold one streamed event into this observation (matches the Go runStreaming aggregation).</summary>
    public void Absorb(StreamResponse ev)
    {
        switch (ev.PayloadCase)
        {
            case StreamResponseCase.Task when ev.Task is not null:
                State = StateToken(ev.Task.Status?.State);
                var (t, present) = TaskArtifactText(ev.Task);
                if (present) { Text += t; ArtifactPresent = true; }
                break;
            case StreamResponseCase.StatusUpdate when ev.StatusUpdate is not null:
                State = StateToken(ev.StatusUpdate.Status?.State);
                break;
            case StreamResponseCase.ArtifactUpdate when ev.ArtifactUpdate?.Artifact is not null:
                foreach (var part in ev.ArtifactUpdate.Artifact.Parts ?? [])
                    if (part.ContentCase == PartContentCase.Text) { Text += part.Text; ArtifactPresent = true; }
                break;
            case StreamResponseCase.Message when ev.Message is not null:
                Kind = "message";
                Text += MessageText(ev.Message);
                break;
        }
    }

    private static string MessageText(Message? msg)
    {
        if (msg?.Parts is null) return "";
        var sb = new StringBuilder();
        foreach (var part in msg.Parts)
            if (part.ContentCase == PartContentCase.Text) sb.Append(part.Text);
        return sb.ToString();
    }

    private static (string Text, bool Present) TaskArtifactText(AgentTask task)
    {
        var sb = new StringBuilder();
        bool present = false;
        foreach (var artifact in task.Artifacts ?? [])
            foreach (var part in artifact.Parts ?? [])
                if (part.ContentCase == PartContentCase.Text) { sb.Append(part.Text); present = true; }
        return (sb.ToString(), present);
    }

    /// <summary>
    /// Map the .NET TaskState enum to the canonical proto token a2a-go/python emit
    /// (e.g. InputRequired -> TASK_STATE_INPUT_REQUIRED), so the harness assertions match.
    /// </summary>
    private static string StateToken(TaskState? state)
    {
        if (state is null || state == TaskState.Unspecified) return "";
        var sb = new StringBuilder("TASK_STATE_");
        var name = state.ToString()!; // PascalCase, e.g. "InputRequired"
        for (int i = 0; i < name.Length; i++)
        {
            if (i > 0 && char.IsUpper(name[i])) sb.Append('_');
            sb.Append(char.ToUpperInvariant(name[i]));
        }
        return sb.ToString();
    }
}

File: integrations/agntcy-a2a-slimrpc/fixtures/node/src/observation.ts
// Copyright AGNTCY Contributors (https://github.com/agntcy)
// SPDX-License-Identifier: Apache-2.0

import {
  type Artifact,
  type Message,
  type StreamResponse,
  type Task,
  type TaskState,
  taskStateToJSON,
} from '@a2a-js/sdk';

/**
 * Parseable view of a SendMessage result, mirroring the Go/Python/.NET/Java probes.
 * Emitted as CSIT_SLIM_* KEY=value lines consumed by matrix_test.go.
 */
export interface Observation {
  kind: string; // "task" | "message" | "unknown"
  state: string; // TASK_STATE_* token; empty for a bare message
  artifactPresent: boolean;
  text: string;
  streamEvents: number;
}

/** Normalizes a proto TaskState (enum number/undefined) to its TASK_STATE_* token, "" if unset. */
export function stateToken(state: TaskState | undefined): string {
  if (state === undefined || state === 0) {
    return '';
  }
  return taskStateToJSON(state);
}

function aggregateArtifacts(artifacts: Artifact[] | undefined): [string, boolean] {
  let text = '';
  let present = false;
  for (const artifact of artifacts ?? []) {
    for (const part of artifact.parts ?? []) {
      if (part.content?.$case === 'text') {
        text += part.content.value;
        present = true;
      }
    }
  }
  return [text, present];
}

function messageText(message: Message): string {
  let text = '';
  for (const part of message.parts ?? []) {
    if (part.content?.$case === 'text') {
      text += part.content.value;
    }
  }
  return text;
}

/** A non-streaming SendMessage result is a Message or a Task (SendMessageResult). */
export function observeResult(result: Message | Task): Observation {
  if ('messageId' in result && typeof result.messageId === 'string') {
    return {
      kind: 'message',
      state: '',
      artifactPresent: false,
      text: messageText(result),
      streamEvents: 0,
    };
  }
  const task = result as Task;
  const [text, present] = aggregateArtifacts(task.artifacts);
  return {
    kind: 'task',
    state: stateToken(task.status?.state),
    artifactPresent: present,
    text,
    streamEvents: 0,
  };
}

/** Observe a Task returned directly (e.g. from CancelTask). */
export function observeTask(task: Task): Observation {
  const [text, present] = aggregateArtifacts(task.artifacts);
  return {
    kind: 'task',
    state: stateToken(task.status?.state),
    artifactPresent: present,
    text,
    streamEvents: 0,
  };
}

/**
 * Aggregate a streaming SendStreamingMessage response into one observation, counting
 * events (proves the stream was multi-event) and concatenating streamed artifact text
 * in order.
 */
export async function observeStream(
  stream: AsyncGenerator<StreamResponse, void, undefined>,
): Promise<Observation> {
  const obs: Observation = {
    kind: 'task',
    state: '',
    artifactPresent: false,
    text: '',
    streamEvents: 0,
  };
  let buffer = '';
  for await (const response of stream) {
    obs.streamEvents++;
    const payload = response.payload;
    if (!payload) {
      continue;
    }
    switch (payload.$case) {
      case 'task': {
        const task = payload.value;
        obs.state = stateToken(task.status?.state);
        const [text, present] = aggregateArtifacts(task.artifacts);
        if (present) {
          buffer += text;
          obs.artifactPresent = true;
        }
        break;
      }
      case 'statusUpdate':
        obs.state = stateToken(payload.value.status?.state);
        break;
      case 'artifactUpdate': {
        const [text, present] = aggregateArtifacts(
          payload.value.artifact ? [payload.value.artifact] : [],
        );
        if (present) {
          buffer += text;
          obs.artifactPresent = true;
        }
        break;
      }
      case 'message':
        obs.kind = 'message';
        buffer += messageText(payload.value);
        break;
    }
  }
  obs.text = buffer;
  return obs;
}

/** Prints the parseable lifecycle block followed by the raw text (echo substring check). */
export function emit(obs: Observation): void {
  console.log('CSIT_SLIM_RESULT_KIND=' + obs.kind);
  console.log('CSIT_SLIM_TASK_STATE=' + obs.state);
  console.log('CSIT_SLIM_ARTIFACT_PRESENT=' + (obs.artifactPresent ? 'true' : 'false'));
  console.log('CSIT_SLIM_STREAM_EVENTS=' + obs.streamEvents);
  console.log('CSIT_SLIM_ARTIFACT_TEXT=' + obs.text);
  console.log(obs.text);
}
